src/App.js

```diff
@@ -39,44 +51,76 @@ class App extends Component {
 
     e.preventDefault();
 
-    if (username === '') {
+    if (username === "") {
       this.setState({ username: value });
     } else {
-      socket.emit('message', { username, value });
+      socket.emit("message", { username, value });
     }
 
-    this.setState({ value: '' });
+    this.setState({ value: "" });
+  }
+
+  handleTyping() {
+    const { username } = this.state;
+    this.setState({ isTyping: true });
+    socket.emit("user is writing", username);
+    setTimeout(this.stoppedTyping, 4000);
+  }
+
+  stoppedTyping() {
+    this.setState({ isTyping: false });
+    socket.emit("stopped typing message");
   }
 
   render() {
     const { username, value, messages } = this.state;
     let form;
 
-    if (username === '') {
+    if (username === "") {
       form = (
         <form onSubmit={this.handleSubmit}>
           Username:&nbsp;
-          <input ref={input => { this.input = input; }} type="text" value={value} onChange={this.handleChange} />
+          <input
+            ref={input => {
+              this.input = input;
+            }}
+            type="text"
+            value={value}
+            onChange={this.handleChange}
+          />
           <input type="submit" />
         </form>
       );
     } else {
       form = (
         <form onSubmit={this.handleSubmit}>
           <span />
-          <input ref={(input) => { this.input = input; }} type="text" value={value} onChange={this.handleChange} />
+          <input
+            ref={input => {
+              this.input = input;
+            }}
+            type="text"
+            value={value}
+            onChange={this.handleChange}
+            onKeyDown={this.handleTyping}
```

Comment:
Borde kanske använda onChange istället för onKeyDown? Skillnaden är ju att onKeyDown inte reagerar på om man kopierar något och klistrar in med musen.